python/ray/experimental/multiprocessing/pool.py

```diff
@@ -0,0 +1,488 @@
+            self._submit_next_chunk()
+        except queue.Empty:
+            raise TimeoutError
+
+        self._ready_objects.extend(self._result_thread.results()[index])
+        self._next_chunk_index += 1
+
+        return self._ready_objects.pop(0)
+
+
+@ray.remote
+class PoolActor(object):
+    def __init__(self, initializer=None, initargs=None):
+        if initializer:
+            if initargs is None:
+                initargs = ()
+            initializer(*initargs)
+
+    def ping(self):
+        pass
+
+    def run_batch(self, func, batch):
+        results = []
+        for args, kwargs in batch:
+            if args is None:
+                args = tuple()
+            if kwargs is None:
+                kwargs = {}
+            try:
+                results.append(func(*args, **kwargs))
+            except Exception as e:
+                results.append(PoolTaskError(e))
+        return results
+
+
+# https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing.pool
+# TODO(edoakes): cloudpickle can't pickle generator objects.
+class Pool(object):
+    def __init__(self,
+                 processes=None,
+                 initializer=None,
+                 initargs=None,
+                 maxtasksperchild=None):
+        self._closed = False
+        self._initializer = initializer
+        self._initargs = initargs
+        self._maxtasksperchild = maxtasksperchild if maxtasksperchild else -1
+        self._actor_deletion_ids = []
+
+        processes = self._init_ray(processes)
+        self._start_actor_pool(processes)
+
+    def _init_ray(self, processes=None):
+        if not ray.is_initialized():
+            # Cluster mode.
+            if "RAY_ADDRESS" in os.environ:
+                address = os.environ["RAY_ADDRESS"]
+                print("Connecting to ray cluster at address='{}'".format(
+                    address))
+                ray.init(address=address)
+            # Local mode.
+            else:
+                print("Starting local ray cluster")
+                ray.init(num_cpus=processes)
+
+        ray_cpus = int(ray.state.cluster_resources()["CPU"])
+        if processes is None:
+            processes = ray_cpus
+        elif ray_cpus < processes:
+            raise ValueError("Tried to start a pool with {} processes on an "
+                             "existing ray cluster, but there are only {} "
+                             "CPUs in the ray cluster.".format(
+                                 processes, ray_cpus))
+
+        return processes
+
+    def _start_actor_pool(self, processes):
+        self._actor_pool = [self._new_actor_entry() for _ in range(processes)]
+        ray.get([actor.ping.remote() for actor, _ in self._actor_pool])
+
+    def _wait_for_stopping_actors(self, timeout=None):
+        if len(self._actor_deletion_ids) == 0:
+            return
+        if timeout is not None:
+            timeout = float(timeout)
```

Comment:
Why?

Reply:
Ray prints a warning if you pass an integer (we can probably remove this warning now)